File: frontend/lib/admin-meetings/schemas.ts
/**
 * Admin Meetings schemas TypeScript (sub-bloque 7.B.1 FASE 7).
 *
 * Espejo de backend/app/motors/m_meetings/schemas.py. Mantener
 * sincronizado manualmente al cambiar contratos backend.
 *
 * NOTA — H1 audit pre-frontend:
 *   sprint4-types.ts contiene tipos meeting legacy (`MeetingSession`,
 *   `MeetingNotes`, `MeetingLiveOutput`) NO eliminados (otros callers
 *   fuera scope FASE 7 los usan). Aquí definimos shapes nuevos
 *   coherentes con backend ExploratoryMeetingRow ampliado.
 */

// ────────────────────────────────────────────────────────────────────
// Type aliases (sincronizar CHECK constraints migration e8b3c5d70a91)
// ────────────────────────────────────────────────────────────────────

export const MEETING_PLATFORMS = [
  "google_meet",
  "zoom",
  "teams",
  "presencial",
  "jitsi",
  "other",
] as const;
export type MeetingPlatform = (typeof MEETING_PLATFORMS)[number];

export const MEETING_ETAPAS_K = [
  "K.1",
  "K.2",
  "K.3",
  "K.4",
  "K.5",
  "K.6",
  "other",
] as const;
export type MeetingEtapaK = (typeof MEETING_ETAPAS_K)[number];

export const MEETING_STATUSES = [
  "scheduled",
  "in_progress",
  "completed",
  "cancelled",
] as const;
export type MeetingStatus = (typeof MEETING_STATUSES)[number];

export const POST_ACTION_TYPES = [
  "propuesta",
  "create_project",
  "k6_signature",
  "email_summary",
] as const;
export type PostActionType = (typeof POST_ACTION_TYPES)[number];

export const PLATFORM_LABELS: Record<MeetingPlatform, string> = {
  google_meet: "Google Meet",
  zoom: "Zoom",
  teams: "MS Teams",
  presencial: "Presencial",
  jitsi: "Jitsi",
  other: "Otra",
};

export const ETAPA_K_LABELS: Record<MeetingEtapaK, string> = {
  "K.1": "K.1 Lead inicial",
  "K.2": "K.2 Cualificación",
  "K.3": "K.3 Pre-propuesta",
  "K.4": "K.4 Reunión exploratoria",
  "K.5": "K.5 Propuesta enviada",
  "K.6": "K.6 Firma contrato",
  other: "Otra",
};

export const STATUS_LABELS: Record<MeetingStatus, string> = {
  scheduled: "Programada",
  in_progress: "En curso",
  completed: "Completada",
  cancelled: "Cancelada",
};

// ────────────────────────────────────────────────────────────────────
// Inputs (espejo MeetingCreate / Update / Complete / Cancel / PostAction)
// ────────────────────────────────────────────────────────────────────

export interface MeetingCreate {
  client_id: string;
  project_id?: string | null;
  title: string;
  platform?: MeetingPlatform | null;
  meeting_url?: string | null;
  etapa_k?: MeetingEtapaK | null;
  interlocutor_contact_id?: string | null;
  scheduled_at?: string | null; // ISO datetime
  lead_source?: string | null;
}

export interface MeetingUpdate {
  title?: string;
  platform?: MeetingPlatform | null;
  meeting_url?: string | null;
  etapa_k?: MeetingEtapaK | null;
  interlocutor_contact_id?: string | null;
  scheduled_at?: string | null;
  project_id?: string | null;
  notes_markdown?: string | null;
}

export interface MeetingComplete {
  notes_markdown?: string | null;
  duration_minutes?: number | null;
  outputs_agente_18?: Record<string, unknown> | null;
}

export interface MeetingCancel {
  reason?: string | null;
}

export interface MeetingPostActionRequest {
  action_type: PostActionType;
  payload?: Record<string, unknown>;
}

export interface MeetingPostActionResponse {
  action_type: PostActionType;
  success: boolean;
  result?: Record<string, unknown>;
  error_message?: string | null;
}

// ────────────────────────────────────────────────────────────────────
// Outputs
// ────────────────────────────────────────────────────────────────────

export interface MeetingContactSummary {
  contact_id: string;
  full_name: string;
  role_title: string;
  role_category: string;
  email: string;
  notes_excerpt?: string | null;
}

export interface MeetingDetail {
  id: string;
  client_id: string;
  project_id?: string | null;
  title: string;
  platform?: MeetingPlatform | null;
  meeting_url?: string | null;
  etapa_k?: MeetingEtapaK | null;
  interlocutor_contact_id?: string | null;
  interlocutor?: MeetingContactSummary | null;
  meeting_date?: string | null;
  duration_minutes?: number | null;
  status: MeetingStatus;
  completed_at?: string | null;
  cancelled_at?: string | null;
  notes_markdown?: string | null;
  notes_html_sanitized?: string | null;
  sse_session_id?: string | null;
  outputs_agente_18?: Record<string, unknown> | null;
  lead_source?: string | null;
  conversion_status: string;
  proposal_generated_id?: string | null;
  created_at?: string | null;
}

export interface MeetingListItem {
  id: string;
  client_id: string;
  project_id?: string | null;
  title: string;
  platform?: MeetingPlatform | null;
  etapa_k?: MeetingEtapaK | null;
  interlocutor_contact_id?: string | null;
  interlocutor_name?: string | null;
  meeting_date?: string | null;
  status: MeetingStatus;
  duration_minutes?: number | null;
  has_notes: boolean;
}

export interface MeetingSearchResult {
  id: string;
  client_id: string;
  title: string;
  meeting_date?: string | null;
  status: MeetingStatus;
  snippet: string;
}

export interface SSEStartResponse {
  sse_session_id: string;
  stream_url: string;
}

// ────────────────────────────────────────────────────────────────────
// SSE event types — emitidos por POST /agents/18/meeting-update/stream
// ────────────────────────────────────────────────────────────────────

export type SSEProgressPhase = "thinking" | "validating_schema";

export interface SSEEventProgress {
  type: "progress";
  phase: SSEProgressPhase;
  ts?: number;
}

export interface SSEEventInsight {
  type: "insight";
  data: {
    categoria_ens?: string;
    confianza_categoria?: number;
    rationale_categoria?: string;
    madurez_actual?: string;
    horas_marcos_estimadas?: number;
    viabilidad_temporal?: string;
    riesgos_detectados?: Array<Record<string, unknown>>;
    quick_wins_sugeridas?: Array<Record<string, unknown>>;
    preguntas_pendientes?: Array<Record<string, unknown>>;
  };
}

export interface SSEEventDone {
  type: "done";
  latency_ms: number;
  tokens_input?: number | null;
  tokens_output?: number | null;
  schema_valid?: boolean;
  fallback_used?: boolean;
}

export interface SSEEventError {
  type: "error";
  error: string;
}

export type SSEEvent = SSEEventProgress | SSEEventInsight | SSEEventDone | SSEEventError;

// ────────────────────────────────────────────────────────────────────
// Filtros UI inbox
// ────────────────────────────────────────────────────────────────────

export interface MeetingsFilters {
  client_id?: string | null;
  project_id?: string | null;
  status?: MeetingStatus | null;
  limit?: number;
}

File: backend/app/motors/m_meetings/schemas.py
"""Motor Meetings — Schemas Pydantic (sub-bloque 7.A.2).

Espejo del modelo ExploratoryMeetingRow + workflow operations.
Tipos compartidos via Literal (sincronizados con CHECK constraints
migration e8b3c5d70a91).
"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


# ────────────────────────────────────────────────────────────────────
# TYPE ALIASES — sincronizar con migration CHECK constraints
# ────────────────────────────────────────────────────────────────────

MeetingPlatform = Literal[
    "google_meet",
    "zoom",
    "teams",
    "presencial",
    "jitsi",
    "other",
]

MeetingEtapaK = Literal[
    "K.1", "K.2", "K.3", "K.4", "K.5", "K.6", "other",
]

MeetingStatus = Literal[
    "scheduled",
    "in_progress",
    "completed",
    "cancelled",
]

PostActionType = Literal[
    "propuesta",
    "create_project",
    "k6_signature",
    "email_summary",
]


# ────────────────────────────────────────────────────────────────────
# INPUT — create + update
# ────────────────────────────────────────────────────────────────────


class MeetingCreate(BaseModel):
    """Body POST /admin/meetings — create scheduled meeting."""

    client_id: uuid.UUID
    project_id: uuid.UUID | None = None
    title: str = Field(..., min_length=1, max_length=200)
    platform: MeetingPlatform | None = None
    meeting_url: str | None = Field(default=None, max_length=500)
    etapa_k: MeetingEtapaK | None = None
    interlocutor_contact_id: uuid.UUID | None = Field(
        default=None,
        description=(
            "FK opcional client_contacts.id (M30). Si presente, se "
            "loggea interaction='meeting_attended' al complete_meeting."
        ),
    )
    scheduled_at: datetime | None = Field(
        default=None,
        description="Mapped a columna ``meeting_date`` server-side.",
    )
    lead_source: str | None = Field(default=None, max_length=40)


class MeetingUpdate(BaseModel):
    """PATCH partial /admin/meetings/{id} — actualizaciones pre-completion.

    Sub-bloque 7.B.5: ``notes_markdown`` opcional autosave debounce 2s
    desde MeetingNotes frontend. Service render ``notes_html_sanitized``
    server-side al persistir.
    """

    title: str | None = Field(default=None, min_length=1, max_length=200)
    platform: MeetingPlatform | None = None
    meeting_url: str | None = Field(default=None, max_length=500)
    etapa_k: MeetingEtapaK | None = None
    interlocutor_contact_id: uuid.UUID | None = None
    scheduled_at: datetime | None = None
    project_id: uuid.UUID | None = None
    notes_markdown: str | None = Field(default=None, max_length=100_000)


class MeetingComplete(BaseModel):
    """Body POST /admin/meetings/{id}/complete — completar meeting."""

    notes_markdown: str | None = Field(default=None, max_length=100_000)
    duration_minutes: int | None = Field(default=None, ge=0, le=1440)
    outputs_agente_18: dict[str, Any] | None = Field(
        default=None,
        description="Output JSON A18 acumulado durante la reunión.",
    )


class MeetingCancel(BaseModel):
    """Body POST /admin/meetings/{id}/cancel."""

    reason: str | None = Field(default=None, max_length=500)


# ────────────────────────────────────────────────────────────────────
# POST-MEETING ACTIONS — 4 cross-motor
# ────────────────────────────────────────────────────────────────────


class MeetingPostActionRequest(BaseModel):
    """Body POST /admin/meetings/{id}/post-action.

    action_type:
      - ``propuesta`` → A19 RedactorPropuestasAgent → P-001 DOCX
      - ``create_project`` → Core projects service.create_project
      - ``k6_signature`` → M12 magic link FIRMA_DOCUMENTO
      - ``email_summary`` → EmailSender consolidado (template HTML)
    """

    action_type: PostActionType
    payload: dict[str, Any] = Field(default_factory=dict)


class MeetingPostActionResponse(BaseModel):
    """Response post-action genérico (cada motor devuelve su payload)."""

    action_type: PostActionType
    success: bool
    result: dict[str, Any] = Field(default_factory=dict)
    error_message: str | None = None


# ────────────────────────────────────────────────────────────────────
# OUTPUT — detail + list views
# ────────────────────────────────────────────────────────────────────


class MeetingContactSummary(BaseModel):
    """Mini-card interlocutor M30 (cuando interlocutor_contact_id presente)."""

    contact_id: uuid.UUID
    full_name: str
    role_title: str
    role_category: str
    email: str
    notes_excerpt: str | None = Field(
        default=None,
        description="Primeros 200 chars de notes_marcos del contacto.",
    )


class MeetingDetail(BaseModel):
    """Vista detalle meeting con interlocutor mini-card."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    client_id: uuid.UUID
    project_id: uuid.UUID | None = None
    title: str
    platform: MeetingPlatform | None = None
    meeting_url: str | None = None
    etapa_k: MeetingEtapaK | None = None
    interlocutor_contact_id: uuid.UUID | None = None
    interlocutor: MeetingContactSummary | None = None

    meeting_date: datetime | None = None
    duration_minutes: int | None = None
    status: MeetingStatus
    completed_at: datetime | None = None
    cancelled_at: datetime | None = None

    notes_markdown: str | None = None
    notes_html_sanitized: str | None = None

    sse_session_id: uuid.UUID | None = None
    outputs_agente_18: dict[str, Any] | None = Field(
        default=None,
        validation_alias="outputs_agente_18_jsonb",
    )

    lead_source: str | None = None
    conversion_status: str
    proposal_generated_id: uuid.UUID | None = None

    created_at: datetime | None = None


class MeetingListItem(BaseModel):
    """Compact view para listings."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    client_id: uuid.UUID
    project_id: uuid.UUID | None = None
    title: str
    platform: MeetingPlatform | None = None
    etapa_k: MeetingEtapaK | None = None
    interlocutor_contact_id: uuid.UUID | None = None
    interlocutor_name: str | None = Field(
        default=None,
        description="Resolved JOIN client_contacts.full_name.",
    )
    meeting_date: datetime | None = None
    status: MeetingStatus
    duration_minutes: int | None = None
    has_notes: bool = False


class MeetingSearchResult(BaseModel):
    """Result item de full-text search GIN sobre notes_markdown."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    client_id: uuid.UUID
    title: str
    meeting_date: datetime | None = None
    status: MeetingStatus
    snippet: str = Field(
        description="Primeros 200 chars notes_markdown que matcheó.",
    )


# ────────────────────────────────────────────────────────────────────
# SSE — A18 stream
# ────────────────────────────────────────────────────────────────────


class SSEStartResponse(BaseModel):
    """POST /admin/meetings/{id}/sse-init crea sse_session_id."""

    sse_session_id: uuid.UUID
    stream_url: str = Field(
        description=(
            "GET /api/v1/agents/18/meeting-update/stream/{sse_session_id} "
            "para consumer EventSource frontend."
        ),
    )
